File: content/snippets/python/design_patterns/facade.py
class Compiler:
    def run(self) -> None:
        print("compile")


class TestRunner:
    def run(self) -> None:
        print("test")


class Packager:
    def run(self) -> None:
        print("package")


class ReleaseFacade:
    def __init__(self) -> None:
        self.compiler = Compiler()
        self.tests = TestRunner()
        self.packager = Packager()

    def deploy(self) -> None:
        self.compiler.run()
        self.tests.run()
        self.packager.run()


ReleaseFacade().deploy()

File: content/snippets/scala2/design_patterns/facade.scala
// Facade
class Compiler {
    def compile(): Unit = println("Compiling code...")
}

class Tester {
    def test(): Unit = println("Running tests...")
}

class Deployer {
    def deploy(): Unit = println("Deploying package...")
}

class BuildFacade {
    private val compiler = new Compiler()
    private val tester = new Tester()
    private val deployer = new Deployer()
    
    def build(): Unit = {
        compiler.compile()
        tester.test()
        deployer.deploy()
    }
}

object Main extends App {
    val facade = new BuildFacade()
    facade.build()
}
